tests/test_client.py

```diff
@@ -432,6 +432,76 @@ def test_json_request(self):
         self.assertEqual(response.headers['Content-type'], 'application/json')
         self.assertIn('Authorization', response.headers)
 
+class TestRequestsV4(unittest.TestCase):
+    # usful args for testing GETs
+    args_in = {
+        'foo':['bar'],
+        'baz':['qux', 'quux=quuux', 'foobar=foobar&barbaz=barbaz']}
+    args_out = dict(
+        (key, [v for v in val])
+        for (key, val) in list(args_in.items()))
+
+    def setUp(self):
+        self.client = duo_client.client.Client(
+            'test_ikey', 'test_akey', 'example.com', sig_timezone='America/Detroit',
+            digestmod=hashlib.sha512, sig_version=4)
+        # monkeypatch client's _connect()
+        self.client._connect = lambda: util.MockHTTPConnection()
+
+    def test_get_no_params(self):
+        (response, dummy) = self.client.api_call('GET', '/foo/bar', {})
+        self.assertEqual(response.method, 'GET')
+        self.assertEqual(response.uri, '/foo/bar?')
+
+    def test_get_params(self):
+        (response, dummy) = self.client.api_call(
+            'GET', '/foo/bar', self.args_in)
+        self.assertEqual(response.method, 'GET')
+        (uri, args) = response.uri.split('?')
+        self.assertEqual(uri, '/foo/bar')
+        self.assertEqual(util.params_to_dict(args), self.args_out)
+
+    def test_json_api_call_get_no_params(self):
+        response = self.client.json_api_call('GET', '/foo/bar', {})
+        self.assertEqual(response['method'], 'GET')
+        self.assertEqual(response['uri'], '/foo/bar?')
+        self.assertEqual(response['body'], None)
+
+    def test_json_api_call_get_params(self):
+        response = self.client.json_api_call(
+            'GET', '/foo/bar', self.args_in)
+        self.assertEqual(response['method'], 'GET')
+        (uri, args) = response['uri'].split('?')
+        self.assertEqual(uri, '/foo/bar')
+        self.assertEqual(util.params_to_dict(args), self.args_out)
+
+    def test_json_post(self):
+        (response, dummy) = self.client.api_call('POST', '/foo/bar', JSON_BODY)
+
+        self.assertEqual(response.method, 'POST')
+        self.assertEqual(response.uri, '/foo/bar')
+        self.assertEqual(response.body, JSON_STRING)
+
+        self.assertIn('Content-type', response.headers)
+        self.assertEqual(response.headers['Content-type'], 'application/json')
+        self.assertIn('Authorization', response.headers)
```

Comment:
Why are you explicitly testing for the existence of an authorization header here?  or alternately, why not test for the authn header in the GET tests above?  both kinds of requests have the header IIRC?

Reply:
To be honest, I copied this entire test class from `TestJsonRequests`, which is all sig v3, and added some cases from `TestRequests` to ensure all these cases work for sig v4. I'll add the header assertions to all the cases.